core/consensus/babe/impl/babe_impl.cpp

```diff
@@ -0,0 +1,231 @@
+        error_channel_{event_bus_.getChannel<event::BabeErrorChannel>()} {
+    BOOST_ASSERT(lottery_);
+    BOOST_ASSERT(proposer_);
+    BOOST_ASSERT(block_tree_);
+    BOOST_ASSERT(network_);
+    BOOST_ASSERT(clock_);
+    BOOST_ASSERT(hasher_);
+    BOOST_ASSERT(log_);
+  }
+
+  void BabeImpl::runEpoch(Epoch epoch,
+                          BabeTimePoint starting_slot_finish_time) {
+    BOOST_ASSERT(!epoch.authorities.empty());
+    log_->info("starting an epoch with index {}", epoch.epoch_index);
+
+    current_epoch_ = std::move(epoch);
+    current_slot_ = current_epoch_.start_slot;
+    slots_leadership_ = lottery_->slotsLeadership(current_epoch_, keypair_);
+    next_slot_finish_time_ = starting_slot_finish_time;
+
+    runSlot();
+  }
+
+  BabeMeta BabeImpl::getBabeMeta() const {
+    return BabeMeta{current_epoch_,
+                    current_slot_,
+                    slots_leadership_,
+                    next_slot_finish_time_};
+  }
+
+  void BabeImpl::runSlot() {
+    using std::chrono::operator""ms;
+    static constexpr auto kMaxLatency = 5000ms;  // just a value
+
+    if (current_slot_ == current_epoch_.epoch_duration) {
+      // end of the epoch
+      return finishEpoch();
+    }
+    log_->info("starting a slot with number {}", current_slot_);
+
+    // check that we are really in the middle of the slot, as expected; we can
+    // cooperate with a relatively little (kMaxLatency) latency, as our node
+    // will be able to retrieve
+    auto now = clock_->now();
+    if (now > next_slot_finish_time_
+        && (now - next_slot_finish_time_ > kMaxLatency)) {
+      // we are too far behind; after skipping some slots (or even epochs)
+      // control will be returned to this method
+      return synchronizeSlots();
+    }
+
+    // everything is OK: wait for the end of the slot
+    timer_.expires_at(next_slot_finish_time_);
+    timer_.async_wait([this](auto &&ec) {
+      if (ec) {
+        log_->error("error happened while waiting on the timer: {}",
+                    ec.message());
+        return error_channel_.publish(Error::TIMER_ERROR);
+      }
+      finishSlot();
+    });
+  }
+
+  void BabeImpl::finishSlot() {
+    auto slot_leadership = slots_leadership_[current_slot_];
+    if (slot_leadership) {
+      processSlotLeadership(*slot_leadership);
+    }
+
+    ++current_slot_;
+    next_slot_finish_time_ += current_epoch_.slot_duration;
+    runSlot();
+  }
+
+  void BabeImpl::processSlotLeadership(const crypto::VRFOutput &proof) {
+    // build a block to be announced
+    const auto &best_block_hash = block_tree_->deepestLeaf();
+
+    BabeBlockHeader babe_header{current_slot_, proof, authority_id_};
+    auto encoded_header_res = scale::encode(babe_header);
+    if (!encoded_header_res) {
+      return log_->error("cannot encode BabeBlockHeader: {}",
+                         encoded_header_res.error().message());
+    }
+
+    // TODO(akvinikym): understand, where to take InherentData from
```

Comment:
Inherent data is described in `Block Inherent Data` section in spec. (Currently 3.3.1.3). Please either create a task to implement that or add it to this PR

Reply:
Maybe this helps
https://substrate.dev/docs/en/overview/extrinsics#inherent-extrinsics-aka-inherents